posts/multiple-filters.md

```diff
@@ -0,0 +1,449 @@
+
+As an example, `availableTransactionsByReceipt` is now:
+
+```ts
+{
+  all: [
+    { text: 'Coop', category: 'food', receipt: 'missing', year: '2018' },
+    { text: 'Tesco', category: 'food', receipt: 'present', year: '2019' },
+  ],
+  present: [
+    { text: 'Tesco', category: 'food', receipt: 'present', year: '2019' },
+  ],
+  missing: [
+    { text: 'Coop', category: 'food', receipt: 'missing', year: '2018' },
+  ],
+}
+```
+
+### Abstracting the filter application
+
+Applying every permutation of all filters except one gets tedious, and is prone to errors. Even though we refrained from writing abstractions previously, it prevents quite a bit of overhead in this case. Also, the domain should not be able to affect the assumptions we have to make to apply these filters (namely, that the grouping of transactions by filter configuration is an object with keys corresponding to the legal filter configurations, and values representing the transactions available through that configuration), so we should be safe.
+
+We can introduce functions to apply each filter, and then another function to apply all filters except one:
+
+```ts
+type FilterName = keyof Filters;
+
+type FilterFunctions = {
+  [name in FilterName]: (
+    transactions: Transaction[],
+    configurations: Filters,
+  ) => Transaction[]
+};
+
+const filterFunctions: FilterFunctions = {
+  category(transactions: Transaction[], configurations: Filters) {
+    return categoryCounts(transactions)[configurations.category];
+  },
+
+  receipt(transactions: Transaction[], configurations: Filters) {
+    return receiptCounts(transactions)[configurations.receipt];
+  },
+
+  year(transactions: Transaction[], configurations: Filters) {
+    return yearCounts(transactions)[filters.year];
+  },
+};
+
+function applyFiltersExcept(
+  except: FilterName,
+  transactions: Transaction[],
+  configurations: Filters,
+) {
+  // Stick all the filter functions in a new array and delete the exception
+  const filterFunctionsWithoutException = {
+    ...filterFunctions(configurations),
+  };
+  delete filterFunctionsWithoutException[except];
+
+  // Get the other filters as an array
+  const otherFilters = Object.values(filterFunctionsWithoutException);
+
+  // Apply all other filters, resulting in only the remaining transactions
+  return otherFilters.reduce((filtered, filter) => filter(filtered), [
+    ...transactions,
+  ]);
+}
+```
+
+We can then use `applyFiltersExcept` like this:
+
+```ts
+const categories = groupByCategory(
+  applyFiltersExcept('category', transactions, filters),
+);
+const receipts = groupByReceipt(
+  applyFiltersExcept('receipt', transactions, filters),
+);
+const years = groupByYear(applyFiltersExcept('year', transactions, filters));
+```
+
+Note: We could use closures to our advantage to avoid passing in transactions and filters to every function call, although that might make the code less readable.
```

Comment:
Maybe we should do this instead. Are we just assuming that readers aren't familiar with closures?